Nicola

NicolaRR512 opened the pull request.

NicolaRR512 committed:
Finished word class

I tried to make the Word class have some of the functionality we need to parse the input correctly

NicolaRR512 committed:
work on parseInput

I've mostly worked on some helper methods for parseInput and added things to the data class that I needed. I also imported the java.lang package

NicolaRR512 committed:
fixed off by one error

wordCount needs to start at 1 since the object has already been created

NicolaRR512 committed:
debugged Word class and checkIfInArray

These seem to be working now

NicolaRR512 committed:
changed function name to be more accurate

NicolaRR512 committed:
debugged isValidNumber

NicolaRR512 committed:
updated isValidNumber

Now it will still count as a valid number if there is a long group of '-' before a number. Which Pyzdrowski said is how it should work in the video

NicolaRR512 committed:
Work on the StringTokenizer

We do need to treat every character other then letters, digits and '-', '\'' as a delimitator. I just rewatched that part of the video. I added every escape character and every special character on my keyboard. He can't expect us to check for every Unicode character can he?

NicolaRR512 committed:
wrote a function to process numeric strings

This function should break it up if the later part is a valid word. I also added some stuff to call the functions. If the stuff I just added but haven't tested works then I think this should function properly. might need to add a case for if the string is empty

lavender-aa commented:
whenever you want me to make a review you can add a comment onto this discussion page. 

also, don't forget to merge main every once in a while; your branch is still 8 commits behind main

lavender-aa commented:
actually, i'll do this instead: 
because you opened this while you weren't done, please open this again when you think everything is done and working properly. when you open a new PR, i'll make a review

lavender-aa commented:
reopening so i can make a review

lavender-aa commented on program-2/Program2.java:
```diff
@@ -76,8 +77,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+            try
+            {
+                bufRead = new BufferedReader( new FileReader(inFile));
+                while(thisLine = bufRead.readLine() != null)
```
compiler error: `Type mismatch: cannot convert from String to boolean`

don't try to assign and check at the same time. do the assignment before, and then check in the while condition.

also, i could be wrong, but it seems like you only ever get the first line from the file (`thisLine` is never updated inside the loop). that might be what you're trying to do inside the condition, but that doesn't work

lavender-aa commented on program-2/Program2.java:
```diff
@@ -76,8 +77,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+        }
+
+        // tokenizes string passed to it and processes tokens
+        static void getTokens(Sting wholeLine, Data data)
```
compiler error: `Sting cannot be resolved to a type`

(fix typo)

lavender-aa commented on program-2/Program2.java:
```diff
@@ -76,8 +77,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+                    while(aToken.charAt(0) == '\'' || aToken.charAt(0) == '-')
+                    {
+                        if (aToken.charAt(0) == '-' && Character.isDigit(aToken.charAt(1))) 
+                            wasNegative = true;
```
compiler error: `wasNegative cannot be resolved to a variable

seems like you may have mixed up a variable name? declaring `wasNegative` and removing `isNegative` seems to work, though  i've only done a bit of testing

NicolaRR512 committed:
fixed conflicts

lavender-aa commented on program-2/Program2.java:
```diff
@@ -76,8 +77,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+                if (Character.isLetter(firstChar)) 
+                {
+                    Word[] wList = data.get_list();
+                    findOrAdd(wList, aToken, data);
```
compiler error: `The method findOrAdd(Word[], Word, Data) in the type Program2 is not applicable for the arguments (Word[], String, Data)`

you're calling `findOrAdd()` with a `String` instead of a `Word`. it seems like you intended it to be a `String`, based on how you call it (multiple times in this program)

lavender-aa commented on program-2/Program2.java:
```diff
@@ -76,8 +77,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+        static boolean isValidNumber(String maybeNumber)
+        {
+            boolean isValid = false;
+            boolean wasNegative = false
```
compiler error: `Syntax error, insert ";" to complete LocalVariableDeclarationStatement`

lavender-aa committed:
resolved Word class merge conflicts

lavender-aa commented on program-2/Program2.java:
```diff
@@ -132,8 +133,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+                    getTokens(thisLine, data);
+                }
+            }
+            catch (IOException error)
```
`FileReader` constructor throws `FileNotFoundException`, add another catch block.

also, the rubric requires that all errors print a message (ex. "Error reading line from input file.")

lavender-aa replied:
this one is my fault for assuming; apparently java will use the IOException block regardless of what kind of error is thrown (given there is at least one thing that throws an IOException)

lavender-aa commented on program-2/Program2.java:
```diff
@@ -190,30 +318,31 @@ static void print_info(Data data) { // asignee: lavender
     int number of occurrences DONE
```
can't make a comment on this since it's not part of your changes, but line 303 (`writer.println("Sum of integers: " + data.get_sum());`) should be adjusted to the correct sum function (`getTotal()`)

lavender-aa commented on program-2/TestNumber.java:
don't include non-project files in your commits; this file was made while you were testing, and was committed

lavender-aa replied:
i actually don't know how to do this via github desktop. i don't imagine it's super hard, but if you need help i can do it for you

NicolaRR512 replied:
This is an accident it keeps automatically selecting them. I should move them to another folder or something.

lavender-aa replied:
since they've already been committed they'll stay on here until deleted. not sure if GHD has an equivalent to `git rm --cached [item]` but that would do the trick

you can just remove them once everything else is working properly

lavender-aa commented on program-2/TestWord.java:
same as other file; don't include

lavender-aa commented on program-2/Program2.java:
```diff
@@ -190,30 +318,31 @@ static void print_info(Data data) { // asignee: lavender
+    private String word;
     private int wordCount;
     private int wordIndex;
+    private boolean isUnique;
```
`wordIndex` and `isUnique` are unnecessary -- the last-used index of the word list exists in the `Data` class, and any new word that is created should be unique anyways (making `isUnique` redundant)

make sure to also remove get/set functions for any of these (if they exist) and remove them from any other functions they're used in

lavender-aa commented on program-2/Program2.java:
```diff
@@ -132,8 +133,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+
+        // loops through passed Word array and checks if Word passed is in it
+        // if not adds the string to the next open spot in the array
+        static void findOrAdd(Word[] wList, Word rightWord, Data data)
```
assuming this function is supposed to take a `String` (see above comment), adjust accordingly

lavender-aa reviewed (changes requested):
this might not be everything, but once the stuff here is fixed i'll make another review where i can hopefully test more without running into compiler errors

lavender-aa commented:
in a separate branch that i don't intend on publishing, i've fixed all the compiler errors and will try and find any other runtime errors/logic errors, review to come

lavender-aa commented on program-2/Program2.java:
```diff
@@ -132,8 +133,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+        static void parse_input(Data data) 
+        { // asignee: nicola 
+            File inFile = data.get_in_file();
+            Word[] wordList = data.get_list();
```
`wordList` is never used (can remove)

lavender-aa replied:
i guess this one isn't an error but reducing warnings is also helpful

lavender-aa commented on program-2/Program2.java:
```diff
@@ -132,8 +133,135 @@ static void get_valid_output(Data data, String[] args) { // asignee: camron
+            // it's a little weird to use a while loop for this, but
+            // this should make it quit early if it finds a match without 
+            // using break
+            while(isNewWord)
```
possible index out of bounds error; can just add a check inside the while here:
`while(isNewWord && count < wList.length)`

NicolaRR512 replied:
I resolved this a different way before, but I changed the integer literal 100 to wList.length() then realized the mistake and changed it to wList.length

lavender-aa reviewed (changes requested):
found some spots with logical/runtime errors. if you try to run Project2 after fixing these you might run into an index out of bound error in one of my parts, you can just comment out whatever is causing it (i'll be fixing it in my branch shortly)

NicolaRR512 committed:
forgot to check if it is a number

I think my Boolean function should work as error checking, but I put it in a try statement anyway just to be extra safe. We can remove it if you think it's unnecessary

NicolaRR512 committed:
Merge branch 'main' into Nicola

NicolaRR512 committed:
Fixed all the compile errors

This version compiles, all the typos are fixed, but it runs with errors. It has an exception when I tried to enter an input file name that I thought would work

lavender-aa commented:
whenever you address an issue from a review you can click the "resolve conversation" button to make it go away. i did it for all the compiler errors, since i've seen that they're gone (but it would take me a while to figure out what else you've addressed so far)

NicolaRR512 committed:
fixed some runtime errors

159 and 194 -added throws IOException two places

211 -used the StringTokenizer to take letters off the end of a numeric string this is an inverse of using it to remove the digits several lines above

172 -added a check to see if a string is empty before it looks for a char so short circuit evaluation stops an index out of bounds error

NicolaRR512 committed:
fixed incrementing last index

I changed a bunch of stuff to get the output file working
We should only use the line I added at 264

NicolaRR512 committed:
changed everything back except corrected line

NicolaRR512 committed:
Another index out of bounds error

This one only trigged if there was exactly one character in the string.
This fixes it by again checking that the length and using the short circuit evaluation

NicolaRR512 committed:
fixed string tokenizer using incorrect delimitor

the string was "123456789"
fixed to "0123456789"

NicolaRR512 committed:
Another out of bounds error

added another if clause so it exits if out of bounds

NicolaRR512 committed:
Update Program2.java

NicolaRR512 committed:
Update Program2.java

NicolaRR512 committed:
Update Program2.java

NicolaRR512 committed:
moved test files deleting them from here

lavender-aa reviewed (approved):
looks good, seems to work with a little bit of testing

lavender-aa commented:
i'm going to merge this quickly so that we can test out the whole program sooner. i think i'll also wait to change the variable and function names until after everything is merged but before we submit
